src/win/util.c

```diff
@@ -107,43 +107,89 @@ int uv_utf8_to_utf16(const char* utf8Buffer, WCHAR* utf16Buffer,
 
 
 int uv_exepath(char* buffer, size_t* size_ptr) {
-  int utf8_len, utf16_buffer_len, utf16_len;
-  WCHAR* utf16_buffer;
+
+  /* According to MSDN the maximum (unicode) length of an NT-style file path */
+  /* is 32767 characters, excluding the null character. Apparently it can be */
+  /* a bit larger due to \\?\, but this probably won't be an issue here. */
+  WCHAR full_path[32768];
+  DWORD result_len;
+  int utf8_len;
+  HANDLE exe_handle;
   int err;
 
   if (buffer == NULL || size_ptr == NULL || *size_ptr == 0) {
     return UV_EINVAL;
   }
 
-  if (*size_ptr > 32768) {
-    /* Windows paths can never be longer than this. */
-    utf16_buffer_len = 32768;
-  } else {
-    utf16_buffer_len = (int) *size_ptr;
+  /* We're going to be working on lengths of at most 32768 characters anyway, */
+  /* so truncate the input size to that length. Note that since size_t is */
+  /* guaranteed to be able to hold values 0 through 65535, this is safe. */
+  if (*size_ptr > ARRAY_SIZE(full_path)) {
+      *size_ptr = ARRAY_SIZE(full_path);
   }
 
-  utf16_buffer = (WCHAR*) uv__malloc(sizeof(WCHAR) * utf16_buffer_len);
-  if (!utf16_buffer) {
-    return UV_ENOMEM;
+  /* Get the path to the module as UTF-16. Normally this should not fail due */
+  /* to the full_path buffer being too short, so if it actually is, return */
+  /* UV_EIO defensively to indicate an internal error. */
+  result_len = GetModuleFileNameW(NULL, full_path, ARRAY_SIZE(full_path));
+  if (result_len == ARRAY_SIZE(full_path)) {
+    err = UV_EIO;
+    goto error;
+  } else if (result_len == 0) {
+    if (GetLastError() == ERROR_INSUFFICIENT_BUFFER) {
+      err = UV_EIO;
+      goto error;
+    } else {
+      err = GetLastError();
+      goto error;
+    }
   }
 
-  /* Get the path as UTF-16. */
-  utf16_len = GetModuleFileNameW(NULL, utf16_buffer, utf16_buffer_len);
-  if (utf16_len <= 0) {
+  /* The path should be null-terminated on success, but take no chances. */
+  full_path[result_len] = L'\0';
+
+  /* Open the executable at that path in read mode to fully resolve it. */
+  exe_handle = CreateFileW(full_path,
+                           GENERIC_READ,
+                           FILE_SHARE_READ,
```

Comment:
Why didn't you just set all share flags (`FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE`) ?